design/src/rfcs/rfc0017_customizable_client_operations.md

```diff
@@ -0,0 +1,198 @@
+> Status: Accepted
+
+For a summarized list of proposed changes, see the [Changes Checklist](#changes-checklist) section.
+
+SDK customers occasionally need to add additional HTTP headers to requests, and currently,
+the SDK has no easy way to accomplish this. At time of writing, the lower level Smithy
+client has to be used to create an operation, and then the HTTP request augmented on
+that operation type. For example:
+
+```rust
+let input = SomeOperationInput::builder().some_value(5).build()?;
+
+let operation = {
+    let op = input.make_operation(&service_config).await?;
+    let (request, response) = op.into_request_response();
+
+    let request = request.augment(|req, _props| {
+        req.headers_mut().insert(
+            HeaderName::from_static("x-some-header"),
+            HeaderValue::from_static("some-value")
+        );
+        Result::<_, Infallible>::Ok(req)
+    })?;
+
+    Operation::from_parts(request, response)
+};
+
+let response = smithy_client.call(operation).await?;
+```
+
+This approach is both difficult to discover and implement since it requires acquiring
+a Smithy client rather than the generated fluent client, and it's anything but ergonomic.
+
+This RFC proposes an easier way to augment requests that is compatible with the fluent
+client.
+
+Terminology
+-----------
+
+- **Smithy Client**: A `aws_smithy_client::Client<C, M, R>` struct that is responsible for gluing together
+  the connector, middleware, and retry policy.
+- **Fluent Client**: A code generated `Client` that has methods for each service operation on it.
+  A fluent builder is generated alongside it to make construction easier.
+
+Proposal
+--------
+
+The code generated fluent builders returned by the fluent client should have a method added to them,
+similar to `send`, but that returns a customizable request. The customer experience should look as
+follows:
+
+```rust
+let response = client.some_operation()
+    .some_value(5)
+    .customize()
+    .await?
+    .mutate_request(|mut req| {
+        req.headers_mut().insert(
+            HeaderName::from_static("x-some-header"),
+            HeaderValue::from_static("some-value")
+        );
+    })
+    .send()
+    .await?;
+```
+
+This new async `customize` method would return the following:
+
+```rust
+pub struct CustomizableOperation<O, R> {
+    handle: Arc<Handle>,
+    operation: Operation<O, R>,
+}
+
+impl<O, R> CustomizableOperation<O, R> {
+    // Allows for customizing the operation's request
+    fn map_request<E>(
+        mut self,
+        f: impl FnOnce(Request<SdkBody>) -> Result<Request<SdkBody>, E>,
+    ) -> Result<Self, E> {
+        let (request, response) = self.operation.into_request_response();
+        let request = request.augment(|req, _props| f(req))?;
+        self.operation = Operation::from_parts(request, response);
+        Ok(self)
+    }
+
+    // Convenience for `map_request` where infallible direct mutation of request is acceptable
+    fn mutate_request<E>(
```

Comment:
This function doesn't use the type parameter <E> which was declared.

I think this is causing me to be unable to compile due to

```
154 |             .mutate_request(|req| {
    |              ^^^^^^^^^^^^^^ cannot infer type for type parameter `E` declared on the associated function `mutate_request`
```

code:

```
            .mutate_request(|req| {
                if let Some(request_id) = request_id {
                    req.headers_mut().insert(
                        HeaderName::from_static("x-amzn-requestid"),
                        HeaderValue::from_static(request_id),
                    );
                }
            })
```

Reply:
This was fixed in https://github.com/awslabs/smithy-rs/pull/1835, which will be included in the next SDK release.